### PR TITLE
32b mode indexing support

### PR DESCRIPTION
**Pull request by shmsong**
This PR enables use of 32b index type when applicable. 

The check is based on computing the `most positive index` and `most negative index` of input `at::Tensor`'s, and will fall back to 64b mode if the check fails. 

The index mode is cached in `SchedulerEntry` so will be accounted for when checking for heuristic cache hit. 

All index mode when not initiated from `FusionExecutorCache` will default to 64b.

- [x] Go through codegen and resource string to use INDEX_TYPE as much as possible
- [x] Check perf difference
- [x] Rebase after #937

**Most affected benchmark runs**

benchmarked commit point:
before 
```
commit 62f40c91ab4d6ac3eb04c8fbd6745f9aa7ef58ff 
Author: Christian Sarofeen <csarofeen@nvidia.com>
Date:   Wed Jun 9 15:27:02 2021 -0400

    Keep generated indexing consistent. (#937)
```
after 
```
commit b5f872b7182c598af2d4e6cfadd99380a8be8854
Author: shmsong <shisong@umich.edu>
Date:   Thu Jun 10 12:58:00 2021 -0700

    cache index mode
```

```
name : MagicScheduler_fp16_Inner_Reduction___GRAPH/MagicScheduler_fp16_Inner_Reduction/2/134217728/manual_time
kernel us before : 1552
kernel us after : 1316
kernel_us diff : -236
cpu_us diff : -236
kernel % diff : -15.206185567010309



name : MagicScheduler_fp16_Inner_Reduction___GRAPH/MagicScheduler_fp16_Inner_Reduction/2/67108864/manual_time
kernel us before : 776
kernel us after : 659
kernel_us diff : -117
cpu_us diff : -117
kernel % diff : -15.077319587628866



name : MagicScheduler_fp16_Inner_Reduction___GRAPH/MagicScheduler_fp16_Inner_Reduction/2/16777216/manual_time
kernel us before : 196
kernel us after : 168
kernel_us diff : -28
cpu_us diff : -28
kernel % diff : -14.285714285714285



name : MagicScheduler_fp16_Inner_Reduction___GRAPH/MagicScheduler_fp16_Inner_Reduction/2/4194304/manual_time
kernel us before : 53
kernel us after : 46
kernel_us diff : -7
cpu_us diff : -7
kernel % diff : -13.20754716981132



name : MagicScheduler_fp16_Outer_Reduction___GRAPH/MagicScheduler_fp16_Outer_Reduction/4194304/2/manual_time
kernel us before : 33
kernel us after : 29
kernel_us diff : -4
cpu_us diff : -4
kernel % diff : -12.121212121212121

```

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 1]
```diff
@@ -614,7 +633,7 @@ std::vector<at::Tensor> FusionExecutor::runFusion(
     }
   }
 
-  KernelArgumentHolder kernel_arguments;
+  KernelArgumentHolder kernel_arguments(options_.index_mode);
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/executor.cpp:647]: options_ has been stored in `FusionExecutor` when `compileFusion`. 
https://github.com/csarofeen/pytorch/blob/e4b37b857966ada532c41da401b290f018242bad/torch/csrc/jit/codegen/cuda/executor.cpp#L157

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 2]
```diff
@@ -27,6 +27,24 @@ namespace cuda {
 
 int FusionExecutor::fusion_id_counter_ = 0; // NOLINT
 
+namespace {
+
+static const char* defineIndexMode(KernelIndexMode index_mode) {
+  switch (index_mode) {
+    case KernelIndexMode::INT32:
+      return "#define INDEX_TYPE int\n";
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/executor.cpp]: nit: we already have typedefs for some integer types. Why not another typedef?

File: torch/csrc/jit/codegen/cuda/utils.h [hunk 3]
```diff
@@ -7,6 +7,8 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
+enum class KernelIndexMode { INT32, INT64 };
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/utils.h]: nit: Should this be placed in type.h?

**Review by naoyam — APPROVED** [review]
LGTM. Just left some minor comments.

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 2 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/executor.cpp]: I changed it to typedef and I agree that it'd be easier to maintain. It was just that using typedef will require INDEX_TYPE to be defined after the typedefs, so I moved the defs around a little bit.

File: torch/csrc/jit/codegen/cuda/utils.h (hunk 3 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/utils.h]: Moved to type.h thanks! Originally thought that it was not part of the IR, but I guess type.h also makes sense.

**Pull request merged**